x-pack/plugins/stack_alerts/server/alert_types/es_query/alert_type.ts

```diff
@@ -0,0 +1,304 @@
+    // the alert state with the timestamp of the latest hit. In the next execution
+    // of the alert, the latestTimestamp will be used to gate the query in order to
+    // avoid counting a document multiple times.
+
+    let timestamp: string | undefined = previousTimestamp;
+    const filter = timestamp
+      ? {
+          bool: {
+            filter: [
+              parsedQuery.query,
+              {
+                bool: {
+                  must_not: [
+                    { bool: { filter: [{ range: { [params.timeField]: { lte: timestamp } } }] } },
+                  ],
+                },
+              },
+            ],
+          },
+        }
+      : parsedQuery.query;
+
+    const query = buildSortedEventsQuery({
+      index: params.index,
+      from: dateStart,
+      to: dateEnd,
+      filter,
+      size: DEFAULT_MAX_HITS_PER_EXECUTION,
+      sortOrder: 'desc',
+      searchAfterSortId: undefined,
+      timeField: params.timeField,
+      track_total_hits: true,
+    });
+
+    logger.debug(`alert ${ES_QUERY_ID}:${alertId} "${name}" query - ${JSON.stringify(query)}`);
+
+    const searchResult: ESSearchResponse<unknown, {}> = await callCluster('search', query);
+
+    if (searchResult.hits.hits.length > 0) {
+      const numMatches = searchResult.hits.total.value;
+      logger.debug(`alert ${ES_QUERY_ID}:${alertId} "${name}" query has ${numMatches} matches`);
+
+      // apply the alert condition
+      const conditionMet = compareFn(numMatches, params.threshold);
+
+      if (conditionMet) {
+        const humanFn = i18n.translate(
+          'xpack.stackAlerts.esQuery.alertTypeContextConditionsDescription',
+          {
+            defaultMessage: `Number of matching documents is {thresholdComparator} {threshold}`,
+            values: {
+              thresholdComparator: getHumanReadableComparator(params.thresholdComparator),
+              threshold: params.threshold.join(' and '),
+            },
+          }
+        );
+
+        const baseContext: EsQueryAlertActionContext = {
+          date: new Date().toISOString(),
+          value: numMatches,
+          conditions: humanFn,
+          hits: searchResult.hits.hits,
+        };
+
+        const actionContext = addMessages(options, baseContext, params);
+        const alertInstance = options.services.alertInstanceFactory(ConditionMetAlertInstanceId);
+        alertInstance
+          // store the params we would need to recreate the query that led to this alert instance
+          .replaceState({ latestTimestamp: timestamp, dateStart, dateEnd })
+          .scheduleActions(ActionGroupId, actionContext);
+        logger.debug(`scheduled actionGroup: ${JSON.stringify(actionContext)}`);
+
+        // update the timestamp based on the current search results
+        const lastTimestamp = searchResult.hits.hits[0]?.sort;
+        if (lastTimestamp != null && lastTimestamp.length !== 0) {
+          timestamp = lastTimestamp[0];
```

Comment:
curious about `lastTimestamp != null && lastTimestamp.length !== 0` - 
When might this happen? And, perhaps we should use `searchResult.hits.hits.find(...` to ensure we don't miss the correct `timestamp` just because the document in index 0 is invalid? 🤔

Reply:
The sort field is an optional `string[]` on the `ESSearchHit` type, which is the reason for the checks, but valid point about using `.find` so I have updated to do that.